### PR TITLE
104 role management

### PR DESCRIPTION
**Pull request by MuhammadUsama1204**

### CODE REVIEW

File: .byebug_history [hunk 1]
```diff
@@ -1,4 +1,6 @@
 c
+params
+c
```
Comment (tauqeer-work92) [inline .byebug_history]: no need to add this file in PR.

File: Gemfile [hunk 2]
```diff
@@ -53,6 +53,10 @@ group :development, :test do
   gem "debug", platforms: %i[ mri mingw x64_mingw ]
   gem "devise"
   gem "byebug"
+  gem 'rolify'
+  gem 'letter_opener'
+  gem 'mailcatcher'
+
 end
```
Comment (tauqeer-work92) [inline Gemfile:59]: remove extra line after adding gem mailcatcher

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,54 +1,68 @@
...
     end
-
-
     private
         def user_params
-            params.require(:user).permit(:username, :email,:password, :password_confirmation)
+            params.require(:user).permit(:username, :email, role_ids: [])
+        end
+        def authorize_admin
+            if current_user&.users_roles.exists?(role_id: Role.find_by(role: 'Admin').id)
```
Comment (tauqeer-work92) [inline app/controllers/users_controller.rb]: better to use a variable for role

`role = Role.find_by(role: 'Admin')`
 
`if current_user&.users_roles.exists?(role: role)`

Comment (tauqeer-work92) [reply app/controllers/users_controller.rb]: we do not need to have exact id, if there is an association as you can see

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,54 +1,68 @@
...
+                flash[:alert] = "You are not authorized to perform this action 1."
+                redirect_to root_path
+            end
+        end
+        def authorize_staff
+            if current_user&.users_roles.exists?(role_id: Role.find_by(role: 'Staff').id)
```
Comment (tauqeer-work92) [inline app/controllers/users_controller.rb]: same like above

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -1,54 +1,68 @@
...
+                flash[:alert] = "You are not authorized to perform this action 2."
+                redirect_to root_path
+            end
+        end
+    
+        def authorize_customer
+            if current_user&.users_roles.exists?(role_id: Role.find_by(role: 'Customer').id)
```
Comment (tauqeer-work92) [inline app/controllers/users_controller.rb]: same as above

Comment (tauqeer-work92) [inline app/controllers/users_controller.rb]: as you are using the same line
 `current_user&.users_roles.exists?(role_id: Role.find_by(role: 'Customer').id)`
 
 maybe use a method for it, which takes the role as an argument

File: test/models/users_role_test.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
```
Comment (tauqeer-work92) [inline test/models/users_role_test.rb:2]: test files not needed

Comment (MuhammadUsama1204) [reply test/models/users_role_test.rb:2]: sure

**Comment by MuhammadUsama1204** [comment]
Updated

**Pull request merged**